### ISSUE TITLE
Add an option to change the end stream clip

### ISSUE THREAD

**Issue by edmael** [owncast/owncast#1965]
At the end of every stream there's an annoying clip that has nothing to do with the stream itself and seems only an "Owncast-brand-thing". Nothing wrong with having something like that, but there's no point in having no option at all to remove it, here's a couple of reason:

- if I've just ran a chill stream and I close the feed my viewers gets blasted with a not-very-pleasant audio in their ears
- injecting something different from what the streamer wants to actually stream without any control on it seems pretty bad, expecially on a self-hosted-platform.

Please consider adding an option to disable that or customize a stream-ended message.

**Comment by gabek** [comment]
The stream ended clip isn't an ego stroking branding thing. It does what it says on the tin: Lets you know the stream is no longer live.  This is particularly important if you're accessing the stream via a player that's not the Owncast web player (smart tv, mobile app, mpv, vlc, etc). If the stream offered up nothing when the stream wasn't active it would just throw some kind of error in the player. Nobody would want that. At least I don't.

This clip simply says "things are working, but the stream is no longer live" and there's no plans to remove this.

**Comment by edmael** [comment]
Thanks for the clarification, but the sound should really be reconsidered: it can ruin the whole mood of what you've just streamed.

I still think an option to enable/disable this is really needed.

**Comment by vivlim** [comment]
@edmael You can replace offline.ts with one that doesn't have audio on v0.0.11 and later, thanks to https://github.com/owncast/owncast/commit/676aa96f40711a42f85f6cf0d49c4c7c9f6a8ab9.
1. Create a copy of offline.ts that doesn't have any audio.
example: `curl https://github.com/owncast/owncast/raw/develop/static/offline.ts | ffmpeg -i pipe: -c copy -an offline.ts`.
if you're using the official docker container (gabekangas/owncast), use `wget -O - https://github.com/owncast/owncast/raw/develop/static/offline.ts | ffmpeg -i pipe: -c copy -an offline.ts` instead, because the container doesn't have curl, but does have wget.
2. Place your offline.ts at `static/offline.ts`, relative to the `owncast` binary.
The easiest way to accomplish this with the official docker container is to create a directory named `static` that contains your `offline.ts` and bind mount it to `/app/static` inside the container.
3. Restart owncast and check it in your browser. You *will* need to clear your cache if you've ever seen the end stream clip in your browser, as offline.ts is downloaded from the server with a cache-control max-age header of *365 days*.
Instead of clearing the cache, you can also use your browser's debugging tools to disable cache while you are refreshing your owncast page to check whether the new offline.ts is being served. This also means that any of your viewers who have already seen the old offline.ts will continue to see it until they clear their caches or watch from a different device.

**Comment by homeisfar** [comment]
I think at the very least if the default stream ended clip did not play audio that would be nice. I've been chided by my friends for that clip specifically because it plays an extra, unexpected sound.

**Comment by vivlim** [comment]
+1, the sound is like 95% of what motivated me to figure out how to replace it.

**Comment by Fyb3roptik** [comment]
> @edmael You can replace offline.ts with one that doesn't have audio on v0.0.11 and later, thanks to [676aa96](https://github.com/owncast/owncast/commit/676aa96f40711a42f85f6cf0d49c4c7c9f6a8ab9).
> 
> 1. Create a copy of offline.ts that doesn't have any audio.
>    example: `curl https://github.com/owncast/owncast/raw/develop/static/offline.ts | ffmpeg -i pipe: -c copy -an offline.ts`.
>    if you're using the official docker container (gabekangas/owncast), use `wget -O - https://github.com/owncast/owncast/raw/develop/static/offline.ts | ffmpeg -i pipe: -c copy -an offline.ts` instead, because the container doesn't have curl, but does have wget.
> 2. Place your offline.ts at `static/offline.ts`, relative to the `owncast` binary.
>    The easiest way to accomplish this with the official docker container is to create a directory named `static` that contains your `offline.ts` and bind mount it to `/app/static` inside the container.
> 3. Restart owncast and check it in your browser. You _will_ need to clear your cache if you've ever seen the end stream clip in your browser, as offline.ts is downloaded from the server with a cache-control max-age header of _365 days_.
>    Instead of clearing the cache, you can also use your browser's debugging tools to disable cache while you are refreshing your owncast page to check whether the new offline.ts is being served. This also means that any of your viewers who have already seen the old offline.ts will continue to see it until they clear their caches or watch from a different device.

This URL is no longer valid

**Comment by gabek** [comment]
Sorry, it got renamed due to conflicts with a TypeScript compiler. https://github.com/owncast/owncast/blob/develop/static/offline.tsclip

**Comment by Fyb3roptik** [comment]
Do I need to recompile this? I have it in static/offline.ts relative to the binary and am using incognito and still am seeing the old one.

**Comment by vivlim** [comment]
It’s a bummer that the only ‘supported’ way to do this is still to follow some janky instructions I reverse engineered nearly a year ago.

I understand the reasoning in https://github.com/owncast/owncast/issues/1965#issuecomment-1167955513 for why there *should* be an end-of-stream clip, but why does it need to have sound? There’s no way for a streamer to adjust the level of that sound to be appropriate and that has very tangible consequences for viewers who have their volume cranked up to hear a quiet stream, ranging from being unpleasant and startling to actually damaging equipment or hearing.

**Comment by jpjonte** [comment]
@Fyb3roptik no need to recompile, just rename the file to `offline.tsclip` like in the comment before yours.

**Comment by zavants** [comment]
The steps do not work, I have tried and Owncast always overwrites the video with its own version.
(Hetzner, Owncast installed as app)

**Comment by zavants** [comment]
The final sound is very loud and disconcerting, I have used Owncast as experimental in personal projects and in all situations the complaint is always the annoying final clip.

**Comment by zavants** [comment]
The final sound is very loud and disconcerting, I have used Owncas tn experimental projects and in all situations the complaint is always the annoying final clip.

**Comment by brixxnatt** [comment]
> @edmael You can replace offline.ts with one that doesn't have audio on v0.0.11 and later, thanks to [676aa96](https://github.com/owncast/owncast/commit/676aa96f40711a42f85f6cf0d49c4c7c9f6a8ab9).
> 
> 1. Create a copy of offline.ts that doesn't have any audio.
>    example: `curl https://github.com/owncast/owncast/raw/develop/static/offline.ts | ffmpeg -i pipe: -c copy -an offline.ts`.
>    if you're using the official docker container (gabekangas/owncast), use `wget -O - https://github.com/owncast/owncast/raw/develop/static/offline.ts | ffmpeg -i pipe: -c copy -an offline.ts` instead, because the container doesn't have curl, but does have wget.
> 2. Place your offline.ts at `static/offline.ts`, relative to the `owncast` binary.
>    The easiest way to accomplish this with the official docker container is to create a directory named `static` that contains your `offline.ts` and bind mount it to `/app/static` inside the container.
> 3. Restart owncast and check it in your browser. You _will_ need to clear your cache if you've ever seen the end stream clip in your browser, as offline.ts is downloaded from the server with a cache-control max-age header of _365 days_.
>    Instead of clearing the cache, you can also use your browser's debugging tools to disable cache while you are refreshing your owncast page to check whether the new offline.ts is being served. This also means that any of your viewers who have already seen the old offline.ts will continue to see it until they clear their caches or watch from a different device.

VIVLIM please can you help in the commands of this

**Comment by vivlim** [comment]
sorry, i haven’t run owncast in about a year, and am unaffiliated with the project. try asking gabek.

**Comment by gabek** [comment]
@brixxnatt There is no official, easy to use support for this in Owncast. I suggest reading through the above thread, in detail, and using the instructions people have given. People have done their own thing to make it work in the past, so it's likely you can figure it out too.

**Comment by homeisfar** [comment]
@brixxnatt I'm not going to write out the full instructions, but the short version is you need a copy of https://github.com/owncast/owncast/blob/develop/static/offline.tsclip this video. Then you can edit it to your heart's content, for example:

`ffmpeg -i offline.tsclip -c:v libx264 -preset veryslow -crf 30 -an offline.ts`

Note the extension. This is an MPEG transport stream container, not mp4.

This will take the original 1MB end stream clip and cut it down to closer to 100kB and without audio. Then rename your result to offline.tsclip and replace the original. If you're using docker/podman then do as vivlim said and just overlay the original version with yours with the volume mount flag.

Of course, you could use any video clip. You don't have to use the one provided with Owncast. If none of what is being said here makes sense to you it would be very beneficial to play around with the Linux terminal and kind of learn about that, first. It would make every other step of self-hosting software much easier.

**Comment by gabek** [comment]
> Of course, you could use any video clip. You don't have to use the one provided with Owncast. 

It would have to be the correct duration, however. I believe it's 8 seconds? This is one of the reasons why I'm afraid to make this a user-customized thing. I fear people will screw it up and break their streams by using 15 second mp4 files or random other combinations. While I could build tooling and sanity checks around this, it's such a niche thing, and it would require a lot of work to make sure people don't screw it up.

It's not that I'm trying to be a jerk, or that I don't care about people customizing their instance. I want that, absolutely, but this is one thing that is easy to mess up, and difficult to put guards around.

**Comment by vivlim** [comment]
My unsolicited suggestion, echoing what homeisfar said above, would be to remove the audio from the clip but otherwise keep it the same. The sound (and especially the *volume* of the sound) seems to be the primary reason that people want to change it in the first place.

**Comment by zavants** [comment]
We need complete and clear instructions. I have tried several times and it works. It's a real headache.

**Comment by working-name** [comment]
> If you're using docker/podman then do as vivlim said and just overlay the original version with yours with the volume mount flag.

That is no longer the case with the self-contained binary which also packages web interface. You'd have to recompile the whole thing to change one asset.

**Comment by charonfaustinus** [comment]
> @brixxnatt There is no official, easy to use support for this in Owncast. I suggest reading through the above thread, in detail, and using the instructions people have given. People have done their own thing to make it work in the past, so it's likely you can figure it out too.

Not even gonna brother implementing an option to disable it???

**Comment by charonfaustinus** [comment]
> @brixxnatt There is no official, easy to use support for this in Owncast. I suggest reading through the above thread, in detail, and using the instructions people have given. People have done their own thing to make it work in the past, so it's likely you can figure it out too.

Not even gonna implementing an option to disable it???
Not everyone is a programmer, So they aren't going to figure it out like you do.

> It's not that I'm trying to be a jerk, or that I don't care about people customizing their instance. I want that, absolutely, but this is one thing that is easy to mess up, and difficult to put guards around.

Just have an option to disable it, if customizing this can break instance.

**Comment by gabek** [comment]
See response above why I will never remove it. It's important. 

> The stream ended clip isn't an ego stroking branding thing. It does what it says on the tin: Lets you know the stream is no longer live.  This is particularly important if you're accessing the stream via a player that's not the Owncast web player (smart tv, mobile app, mpv, vlc, etc). If the stream offered up nothing when the stream wasn't active it would just throw some kind of error in the player. Nobody would want that. At least I don't.
> 
> 
> 
> This clip simply says "things are working, but the stream is no longer live" and there's no plans to remove this.

**Comment by working-name** [comment]
> See response above why I will never remove it. It's important.
> 
> > The stream ended clip isn't an ego stroking branding thing. It does what it says on the tin: Lets you know the stream is no longer live.  This is particularly important if you're accessing the stream via a player that's not the Owncast web player (smart tv, mobile app, mpv, vlc, etc). If the stream offered up nothing when the stream wasn't active it would just throw some kind of error in the player. Nobody would want that. At least I don't.
> > This clip simply says "things are working, but the stream is no longer live" and there's no plans to remove this.

Giving people a choice to remove it or silence it **is not removing it**, right?

**Comment by gabek** [comment]
I will not add a toggle in Owncast that purposefully breaks the product and makes it more confusing for viewers.

I added my reasons why it shouldn't ever be removed and I'm not sure what's not clear. It serves a real purpose. If it were to be removed it would be a complete mess for players. I will never support that idea and you shouldn't either.

**Comment by working-name** [comment]
I guess my problem is I can't follow the logic of the argument at hand.

If there was a pre-stream looping video animation that would help one figure out they can see the video and have a good audio level, then great, all for it. It would even make sense to say removing it would make it less helpful.

However, given that the video animation is playing at the end, with a disproportionate sound level, it's nothing but jarring. It doesn't help one figure out anything. They will have already heard/seen something on the stream before the animation plays, so it serves no troubleshooting purpose. It brings 0 useful information to the table. It clears nothing.

**Comment by gabek** [comment]
Find a stream that is offline. Open VLC. Put in the url of that stream. If the offline clip didn't exist it would give an error since there is no content to play and everything would look broken.

I keep saying the same thing and either I'm doing a bad job at explaining or nobody wants to listen. But I think I've explained myself.

**Comment by vivlim** [comment]
@gabek, It’s been over a year and you haven’t even *acknowledged* any of the concerns about the audio in the clip which have been mentioned by multiple people. Why?

**Comment by aaronpk** [comment]
So, funny story, I've been seeing these comments in this thread and it's not until this last message that I realized the value of the offline clip. If you plug in the stream URL into a player when it's offline, I agree you definitely don't want to get an error message. But that seems to be a different problem than what people are complaining about here, which is the jarring effect of the offline clip playing immediately after the stream goes from the content to the offline clip.

so maybe a solution is to leave the offline clip there but not play it when the stream ends, only leave it as the clip that plays if someone loads the stream URL when there isn't an active stream?

**Comment by gabek** [comment]
I think you're right. We might be talking past each other here.

I do think the end clip has value at the end of a stream to make it clear the stream has ended on purpose, and it didn't just break. I'm not totally attached to this, but I'd personally prefer it to stay strictly out of clarity. But it's not required like the other use case.

However, and this probably wasn't clear, I don't care about the contents of this clip at all. I haven't been interested in changing it because 99% of people don't care about it, and I'm in that population. I'm not going to dedicate time swapping out working things if I don't care about it one way or the other. I don't think the audio is jarring, and nobody else seems to either, so I don't spend time on it.

I do find it odd that given how passionate this discussion is that nobody has offered to create a replacement. It's simply been "I hate it, change it because I personally have a problem with it. I'll probably hate your replacement too.". 

If somebody wanted to create a replacement with the same specs, and no, or different audio, if it was a pretty ok clip, I'd happily swap it out. I'm not attached to the current one at all. It's there purely to perform a job. It's the job I care about.

**Comment by working-name** [comment]
![image](https://github.com/owncast/owncast/assets/16505052/a9874e11-68ec-4165-85a7-b3a51680f8fb)
Here ya go.
[offline2.zip](https://github.com/owncast/owncast/files/12741307/offline2.zip)

**Comment by vivlim** [comment]
[Last August I shared the ffmpeg command I used to remove audio from the clip](https://github.com/owncast/owncast/issues/1965#issuecomment-1207282414), and *I never said anything about removing the clip*, but here: [offline.tsclip.zip](https://github.com/owncast/owncast/files/12741889/offline.tsclip.zip)

edit: actually I'll just open a PR.

**Comment by charonfaustinus** [comment]
I actually ended up using https://github.com/Hakkin/streamRIP for my custom offline stream instead. When I disconnected by internet drops out or something else, it'll play my custom video on a loop and keep stream online. Then I can just turn it off later via command line.

<blockquote><img src="https://opengraph.githubassets.com/06fd83b79cfb9ce3950bd52a629612bdf1b4a736e8391e03bad57c90cdb5a2df/Hakkin/streamRIP" width="48" align="right"><div><img src="https://github.githubassets.com/favicons/favicon.svg" height="14"> GitHub</div><div><strong><a href="https://github.com/Hakkin/streamRIP">GitHub - Hakkin/streamRIP: Fallback to looping video file if main stream goes offline</a></strong></div><div>Fallback to looping video file if main stream goes offline - GitHub - Hakkin/streamRIP: Fallback to looping video file if main stream goes offline</div></blockquote>

**Comment by gabek** [comment]
I sat down and put in some thought about how I could support this for people *and* not allow people to break things. And I think I came up with a solution. It's not something I can just throw together, but I think it's the best option.

Problem: If we let people replace the video they will use an incompatible video clip that's too long, too high of a bitrate, an unusable codec, or generally just plain unplayable for the use case, causing errors and buffering when a stream ends or when accessing an offline stream. It will be near impossible to troubleshoot, streamers will get pissed off, and viewers will think everything is broken.

Solution: Make a compatible video clip out of an incompatible one.

At launch, Owncast can take whatever offline.mp4 that's supplied, of whatever codec, length, bitrate, whatever and put it through the Owncast transcoding pipeline, spitting out a transport stream file that matches exactly the rest of the stream. In theory, with this approach the offline video clip can be as long as you want because if we use the Owncast transcoding pipeline we'll just split it into the segment lengths that match the rest of your stream.

This is almost how it already works. At launch, it puts the offline clip through our transcoding pipeline and creates a standalone stream out of it. The server-side stitching when dealing with end-of-stream, however, works differently and doesn't go through the transcoder at all. But if we pre-transcode it at launch, then we don't have to worry about doing it while a stream is live and still get a fully compatible clip from whatever might be thrown at it.

I think this could work with some effort.

**Comment by unbelauscht** [comment]
I spent this evening created a custom "stream offline" clip of 3 minutes (as an outro) and wondered why it behaved so poorly, so I went through this entire thread.

Thank you for reconsidering @gabek <3

**Comment by gabek** [comment]
It's [not that I didn't want to support it](https://github.com/owncast/owncast/issues/1965#issuecomment-1721875687), I just hadn't come up with a solution until now that could work.

**Comment by ryanginn** [comment]
any updates?

**Comment by gabek** [comment]
See above for the discussion thread, the reason why this all exists, and the complexity of simply removing it. There was some brainstorming around possible ideas to maybe add some support at some point in the future. This is the current status.

**Comment by sobujbd** [comment]
I have installed **Owncast** with quick installer. Is there any customization guides?

**Comment by gabek** [comment]
There is currently no supported way to customize this.

**Comment by sobujbd** [comment]
I see development version contain **offline-v2.ts** has no sound?

**Comment by gabek** [comment]
That's correct! No sound going forward.

**Comment by anodic-tech** [comment]
I just ran into this, and would like to add that yes the clip has value, but it is also something that is trivial for a streamer to do themselves when they end their stream. I personally would prefer not to have Owncast branding at the end of my stream and would like that as an option.

**Comment by gabek** [comment]
Renamed to reflect the subtasks that were filed.

**Comment by zerofill** [comment]
I know adding a custom video for the docker version to app/static when mapping that volume doesn't work. At least not for me.
